### PR TITLE
Import-DbaCsv - Add support for Azure SQL Database

### LINKED ISSUE #6855: Import-DbaCsv support for Azure SQL DB
#### Host used

- [x] powershell.exe
- [ ] ISE
- [ ] VS Code
- [ ] Other (please specify)

#### Errors Received

```
Exception calling "ChangeDatabase" with "1" argument(s): "USE statement is not supported to switch between databases. Use a new connection to connect to a 
different database."
At C:\Program Files\WindowsPowerShell\Modules\dbatools\1.0.114\allcommands.ps1:38614 char:17
+                 $sqlconn.ChangeDatabase($Database)
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : SqlException
 
Exception calling "ExecuteScalar" with "0" argument(s): "Reference to database and/or server name in '[DB].sys.schemas' is not supported in this version 
of SQL Server."
At C:\Program Files\WindowsPowerShell\Modules\dbatools\1.0.114\allcommands.ps1:38622 char:21
+                 if (($sqlcmd.ExecuteScalar()) -eq 0) {
+                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : SqlException
 
Exception calling "ExecuteScalar" with "0" argument(s): "Reference to database and/or server name in '[DB].sys.tables' is not supported in this version of 
SQL Server."
At C:\Program Files\WindowsPowerShell\Modules\dbatools\1.0.114\allcommands.ps1:38646 char:21
+ ...         if (($sqlcmd.ExecuteScalar()) -eq 0 -and ($sqlcmd2.ExecuteSca ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : SqlException
 
WARNING: [11:55:45][Import-DbaCsv] Failure | The transaction is either not associated with the current connection or has been completed.
```

#### Steps to Reproduce

Import-DbaCsv -SqlInstance $sqlserver -Database $sqldatabase

#### Expected Behavior

To use the db in the setting

#### Actual Behavior

It tried to switch db

### Environmental information

🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨
Please provide the output of the below script

```powershell
PowerShell Version        : 5.1.18362.752
dbatools latest installed : 1.0.114
Culture of OS             : en-GB
```

#### SQL Server:

```sql
Microsoft SQL Azure (RTM) - 12.0.2000.8   Jul 31 2020 08:26:29   Copyright (C) 2019 Microsoft Corporation 
```

```sql
us_english
```

### PR DESCRIPTION
**Pull request by lancasteradam**
- Removed the code that called 'use <db>' since Azure SQL doesn't allow that to be called.
- Updated to use Connect-DbaInstance with the -Database param to set the db connection context. If the database does not exist an error will be returned "cannot open database".

## Type of Change

 - [x] Bug fix (non-breaking change, fixes #6855 )
 - [ ] New feature (non-breaking change, adds functionality, fixes # )
 - [ ] Breaking change (effects multiple commands or functionality, fixes # )
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [ ] Adding code coverage to existing functionality
 - [x] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/sqlcollaborative/appveyor-lab ?
 - [ ] Nunit test is included
 - [ ] Documentation
 - [ ] Build system
 

### Purpose
 
Minor fix to support Azure SQL.

### Commands to test

$serverAzureTest = Connect-DbaInstance -ConnectionString "Data Source=TCP:yourserver.database.windows.net,1433;MultipleActiveResultSets=False;Connect Timeout=30;Encrypt=True;TrustServerCertificate=False;User Id=youruserid;Password=strongpassword;Database=TestDB"

$result = Import-DbaCsv -Path "your sample tab delimited file" -SqlInstance $serverAzureTest -Database TestDB -Delimiter `t -Table TestTable -Schema Test -Truncate -AutoCreateTable -Verbose

### CODE REVIEW

File: functions/Import-DbaCsv.ps1 [hunk 1]
```diff
@@ -439,7 +439,7 @@ function Import-DbaCsv {
                 $elapsed = [System.Diagnostics.Stopwatch]::StartNew()
                 # Open Connection to SQL Server
                 try {
-                    $server = Connect-SqlInstance -SqlInstance $instance -SqlCredential $SqlCredential -StatementTimeout 0 -MinimumVersion 9
+                    $server = Connect-DbaInstance -SqlInstance $instance -SqlCredential $SqlCredential -Database $Database -StatementTimeout 0 -MinimumVersion 9
```
Comment (wsmelton) [inline functions/Import-DbaCsv.ps1:442]: Just add the database parameter to our internal command. We only use the public function on invoke command (there might be a few exclusions).

**Review by wsmelton — CHANGES_REQUESTED** [review]

File: functions/Import-DbaCsv.ps1 (hunk 1 above)
Comment (wsmelton) [reply functions/Import-DbaCsv.ps1:442]: Actually the internal function has the database parameter so please revert this back and include reference to the database parameter.

Comment (wsmelton) [inline functions/Import-DbaCsv.ps1:442]: Please revert to using the `Connect-SqlInstance` command. It has a `-Database` parameter.

I do recall though that we didn't use that in this command for some reason but I've completely forgotten now why.

**Review by wsmelton — CHANGES_REQUESTED** [review]

File: functions/Import-DbaCsv.ps1 (hunk 1 above)
Comment (lancasteradam) [reply functions/Import-DbaCsv.ps1]: Sure, no problem. I switched it back to Connect-SqlInstance. There is a boundary case that I added a check for in Connect-SqlInstance. Maybe that is the scenario that was observed back when the Import-DbaCsv command was being written. It could explain why the Import-DbaCsv command previously had separate code to check if the database existed (rather than letting Connect-SqlInstance/Connect-DbaInstance handle that).  

On a separate note, is there a long term plan to remove Connect-SqlInstance since it just uses Connect-DbaInstance? It might reduce maintenance effort/cost.

btw, Happy New Year!

Comment (wsmelton) [reply functions/Import-DbaCsv.ps1]: No. It stays because it is for we control internal use cases with commands that can hit Azure SQL. There is no reason to put that in a public command. It is just a wrapper now for the public function.

**Comment by wsmelton** [comment]
The internal function is used in all of our commands. Making a change to that will have to be in a PR by itself so we can verify what breaks and what does not.

**Comment by lancasteradam** [comment]
Ok, that makes sense. The change I made to Connect-SqlInstance seems to have affected other commands based on the appveyor results, so I've reverted the change.

I switched the Import-DbaCsv back to using Connect-DbaInstance since it handles the boundary case I ran into. The boundary case is:
1. Connect to a server with one database specified (See the commands under 'Commands to test' above)
2. Re-use that same server object and call Import-DbaCsv with an invalid db name.

When this scenario occurs Import-DbaCsv doesn't catch the invalid db name since Connect-SqlInstance just returns the existing server object.

If there is a strong preference to not use Connect-DbaInstance I can add code to check if the db exists.

**Comment by wsmelton** [comment]
That edge case is irrelevant and will break on many other commands.

**Comment by wsmelton** [comment]
> Connect-SqlInstance just returns the existing server object.

They is why the import function code was manipulating the connection.

I can fully review and test this later, I'm just looking at this from my phone right now so can't get the full picture.

**Comment by lancasteradam** [comment]
Just adding some changes based on the discussion and checking the appveyor results. No hurry on the review since we've got the holiday weekend.

**Comment by lancasteradam** [comment]
@wsmelton, let me know if any additional changes are needed.

**Comment by wsmelton** [comment]
Test results between dev branch and PR:

```powershell
[PS 5.1] [16:28:44] [  9.6 s ][ git:development ] C:\git\dbatools
 [12] > Import-DbaCsv -FullPath C:\temp\testexport.csv -SqlInstance sql -SqlCredential $containerCred -Database tempdb -Table dbatoolslci_testImportdev -Schema dbo -Truncate -AutoCreateTable -NoProgress


ComputerName  : SQL
InstanceName  : MSSQLSERVER
SqlInstance   : SQL
Database      : tempdb
Table         : dbatoolslci_testImportdev
Schema        : dbo
RowsCopied    : 6347
Elapsed       : 1.98 s
RowsPerSecond : 3211
Path          : C:\temp\testexport.csv



[PS 5.1] [16:29:14] [  2.0 s ][ git:development ] C:\git\dbatools
 [13] > remove-module dbatools
[PS 5.1] [16:29:22] [ 45.0ms ][ git:development ] C:\git\dbatools
 [14] > git branch
  development
  ga-repo-folder-restructure
  master
  pr/lancasteradam/7054
* pr/lancasteradam/7064
[PS 5.1] [16:29:45] [842.0ms ][ git:pr/lancasteradam/7064 ] C:\git\dbatools
 [15] > ipmo .\dbatools.psd1 -force
[PS 5.1] [16:30:03] [ 10.4 s ][ git:pr/lancasteradam/7064 ] C:\git\dbatools
 [16] > Import-DbaCsv -FullPath C:\temp\testexport.csv -SqlInstance sql -SqlCredential $containerCred -Database tempdb -Table dbatoolslci_testImportpr -Schema dbo -Truncate -AutoCreateTable -NoProgress


ComputerName  : SQL
InstanceName  : MSSQLSERVER
SqlInstance   : SQL
Database      : tempdb
Table         : dbatoolslci_testImportpr
Schema        : dbo
RowsCopied    : 6347
Elapsed       : 937.56 ms
RowsPerSecond : 6774
Path          : C:\temp\testexport.csv
```

CSV file utilized:
[testexport.txt](https://github.com/sqlcollaborative/dbatools/files/5791928/testexport.txt)

**Comment by wsmelton** [comment]
Test results on the file against Azure SQL:

```powershell
[PS 5.1] [16:18:16] [ 18.1 m ][ git:pr/lancasteradam/7064 ] C:\git\dbatools
 [7] > Import-DbaCsv -FullPath C:\temp\testexport.csv -SqlInstance thy.database.windows.net -SqlCredential (Get-Credential) -Database ss -Table dbatoolslci_testImportpr -Schema dbo -Truncate -AutoCreateTable -NoProgress

cmdlet Get-Credential at command pipeline position 1
Supply values for the following parameters:
Credential

ComputerName  : thy.database.windows.net
SqlInstance   : thy
Database      : ss
Table         : dbatoolslci_testImportpr
Schema        : dbo
RowsCopied    : 6347
Elapsed       : 7.92 s
Path          : C:\temp\testexport.csv
```

**Review by wsmelton — APPROVED** [review]

File: functions/Import-DbaCsv.ps1 [hunk 2]
```diff
@@ -439,7 +439,14 @@ function Import-DbaCsv {
                 $elapsed = [System.Diagnostics.Stopwatch]::StartNew()
                 # Open Connection to SQL Server
                 try {
-                    $server = Connect-SqlInstance -SqlInstance $instance -SqlCredential $SqlCredential -StatementTimeout 0 -MinimumVersion 9
+                    $server = Connect-SqlInstance -SqlInstance $instance -SqlCredential $SqlCredential -Database $Database -StatementTimeout 0 -MinimumVersion 9
+
+                    # boundary case: ensure the correct db is referenced in the $server object
+                    if ($Database -ne $server.ConnectionContext.DatabaseName) {
+                        Stop-Function -Message "The SqlInstance object for $instance is currently connected to the database `"$($server.ConnectionContext.DatabaseName)`" and does not match the parameter `"-Database $Database`". If a pre-connected SqlInstance object is passed in it must be connected to the database specified by the -Database parameter."
```
Comment (wsmelton) [inline functions/Import-DbaCsv.ps1:446]: Connecting to Azure if `Connect-DbaInstance` is used to create a connection object and the database **is not provided** then the database name on that context is null. Which causes the message to not provide the current database name.

A more consistent property and one that will be populated, regardless, is `CurrentDatabase`:

```powershell
[PS 5.1] [16:39:40] [  4.0ms ][ git:pr/lancasteradam/7064 ] C:\git\dbatools
 [27] > $cn.ConnectionContext             


StatementTimeout          : 600
ConnectionType            : DatabaseEngine
BatchSeparator            : GO
TransactionDepth          : 0
SqlExecutionModes         : ExecuteSql    
FixedServerRoles          : None
UserProfile               : 
ProcessID                 : 63
TrueLogin                 : localadmin
TrueName                  : thy
InUse                     : True
LockTimeout               : -1
ServerVersion             : 12.0.2000
ProductVersion            : 12.0.2000.8
DatabaseEngineType        : SqlAzureDatabase
DatabaseEngineEdition     : SqlDatabase
HostPlatform              : Windows
ConnectionProtocol        : TcpIp
IsContainedAuthentication : False
SqlConnectionObject       : System.Data.SqlClient.SqlConnection
CurrentDatabase           : master
IsOpen                    : True
CapturedSql               : Microsoft.SqlServer.Management.Common.CapturedS
AutoDisconnectMode        : DisconnectIfPooled
IsForceDisconnected       : False
AccessToken               : 
ServerInstance            : TCP:thy.database.windows.net,1433
Login                     : localadmin
Password                  : P@ssword$01234
SecurePassword            : System.Security.SecureString
LoginSecure               : False
ConnectAsUserName         : 
ConnectAsUserPassword     : 
ConnectAsUser             : False
NonPooledConnection       : True
PooledConnectionLifetime  : 0
MinPoolSize               : 0
MaxPoolSize               : 100
ConnectTimeout            : 30
Authentication            : NotSpecified
ApplicationIntent         : 
TrustServerCertificate    : False
ConnectionString          : Data Source=TCP:thy.database.windows.net,1433;U
                            PowerShell module - dbatools.io"
SecureConnectionString    : System.Security.SecureString
NetworkProtocol           : NotSpecified
ApplicationName           : 
WorkstationId             : THY-01-0145-LT
DatabaseName              : 
PacketSize                : 8000
MultipleActiveResultSets  : False
EncryptConnection         : False
```

Comment (wsmelton) [reply functions/Import-DbaCsv.ps1:446]:
```powershell
[PS 5.1] [16:41:29] [496.0ms ][ git:pr/lancasteradam/7064 ] C:\git\dbatools
 [32] > Import-DbaCsv -FullPath C:\temp\testexport.txt -SqlInstance $cn -Database ss -Table dbatoolslci_testImportprwithcnobj -Schema dbo -Truncate -AutoCreateTable -NoProgress
WARNING: [16:42:36][Import-DbaCsv] The SqlInstance object for thy.database.windows.net is currently connected to the database "" and does not match the parameter "-Database ss". If a pre-connected SqlInstance object is passed in it must 
be connected to the database specified by the -Database parameter.
```

**Review by wsmelton — CHANGES_REQUESTED** [review]

File: functions/Import-DbaCsv.ps1 (hunk 2 above)
Comment (lancasteradam) [reply functions/Import-DbaCsv.ps1:446]: Thanks for catching that. I see what you mean, even the MS docs have a 'TBD' for the description of the DatabaseName property. The CurrentDatabase property on the other hand does have documentation. I'll upload that change.

**Review by wsmelton — APPROVED** [review]

File: functions/Import-DbaCsv.ps1 [hunk 3]
```diff
@@ -443,7 +443,7 @@ function Import-DbaCsv {
 
                     # boundary case: ensure the correct db is referenced in the $server object
                     if ($Database -ne $server.ConnectionContext.DatabaseName) {
```
Comment (sandscap-sc) [inline functions/Import-DbaCsv.ps1:445]: @lancasteradam In order to keep the logic and error messages in sync, shouldn't we be checking the `CurrentDatabase` property instead of the `DatabaseName` property? Having a `if -ne` check on one property then throwing an error reporting the value of another property feels like it's going to lead to headaches for users.

Comment (lancasteradam) [reply functions/Import-DbaCsv.ps1:445]: @sandcap-sc, great catch, I'll make that change.
